In [12]:
import numpy as np


# Class

class Quadratic:
   
    def __init__(self, coeffs):  #Store coefficients as floats (or complex if provided) as a0 + a1x + a2x^2.
        if len(coeffs) >= 4:
            print("A quadratic requires at most 3 coefficients: [a0, a1, a2].")
        self.coeffs = np.array(coeffs)  #Keep user's numeric types but ensure it's a NumPy array for easy arithmetic
       
# Method 1
    
    def evaluate(self, x):
        result = float(0)
        for i, b in enumerate(self.coeffs):
            result += b*(x**i)
        return result
        
        
# Method 2
        
    def roots(self):
        while len(self.coeffs) > 1 and self.coeffs[-1] == 0:  #This removes unwanted zeros so we can break down findnig the roots into three cases (dependent on number of coefficients).   
            self.coeffs = self.coeffs[:-1]
            
        if len(self.coeffs) == 3:     #The 1st case for 3 coefficients (full quadratic).
           
            a = float(self.coeffs[2])    #Saving the coefficients as there common name for implementation of the quaratic formula.
            b = float(self.coeffs[1])
            c = float(self.coeffs[0])
            
            D = b**2 - 4*a*c
            root_D = np.sqrt(D + 0j)  #This part calculates the discriminant in NumPy so it can save the result as a complex number (if neccesary).
          
            
            root_1 = (-b + root_D)/(2*a)  #Calculates and associates the (possible) two roots to variables.
            root_2 = (-b - root_D)/(2*a)
                      
            if root_1 == root_2:  #Here is the case of a repeated root.
                print("There is a repeated solution to f(x) = 0 at x = ",root_1.real) 
                
            elif D < 0:   #Here is the case that we have complex roots.
                print("The quadratic has two complex solutions to f(x) = 0; at x = ",root_1.real,"+",root_1.imag,"i", ", and x = ",root_2.real,"+",root_2.imag,"i")   #Here using .real and .imag so it prints the results more cleanly.
                
            else:   #Here is the case for two distinct real roots.
                print("The quadratic has two solutions to f(x) = 0; at x = ",root_1.real, ", and x = ",root_2.real)
                
        elif len(self.coeffs) == 2:          #The 2nd case is for solving a linear equation (2 coefficients). 
            root = -(self.coeffs[0])/self.coeffs[1]      #Calculating the root: If ax + b = 0, x = -b/a
            print("This linear equation has a solution to f(x) = 0 at x = ",root)
            
        else:
            if self.coeffs[0] == 0:     #Here is the case for a constant equation; roots only exist if the constant coefficient is 0, as 0 = 0 for all x values
                print("This is a constant equation that can be represented as y = 0, thus it has infinite solutions at all x values.")
                
            else:
                print("This is a constant equation that doesn't intersect the x-axis, thus it has no solutions to f(x) = 0.")
          
            
# Method 3

    def extrema(self, c, d):
        
        c = float(c)  #Saving as float for better calculations.
        d = float(d) 
        
        if d <= c:
            print("It is required that c < d for the closed interval [c, d].")
            
        start = self.evaluate(c)  #Findnig the value of these end points. 
        end = self.evaluate(d)
            
        while len(self.coeffs) > 1 and self.coeffs[-1] == 0:  #Again removing unwanted zeros so we can break down the function into three cases to have a non-zero coeffiient of the highest power of x.   
            
            self.coeffs = self.coeffs[:-1]
                   
        
        if len(self.coeffs) == 3:    #Case 1 again for three coefficients (quadratic with non-zero x-coefficient).
            
            dydx_m = (self.coeffs[2])*2      #Differentiating our quadratic once will give us a linear equation of the form 2(a2)x + a1 (of the form mx + c, with m non-zero) 
            dydx_c = self.coeffs[1]
           
            
            turn_point = float(-(dydx_c)/dydx_m)   #Finding the turning is solving mx + c = 0, giving x = -c/m.
            
            if c <= turn_point <= d: #Subcase for when the turning point is inside the closed interval (including being on the ends of the closed interval).
                
                extreme = self.evaluate(turn_point)  #Finding the value at the turning point.
                
                if self.coeffs[2] > 0:
                    print("This quadratic has a 'postitive' parabola, and the tunrning point is located within the closed interval.") #Different reports based on parabola:
                    print("Thus, the minimum point is the turning point (",turn_point,",",extreme,").")
                    if start == end:
                        print("There are two maximum points at (",c,",",start,"), and (",d,",",end,").") #The case that there are multiple maximum points in the interval.
                    elif max(start, end) == start:
                        print("The maximum point is then (",c,",",start,").")
                    else:
                        print("The maximum point is then (",d,",",end,").")
                        
                elif self.coeffs[2] < 0:
                    print("This quadratic has a 'negative' parabola, and the tunrning point is located within the closed interval.")
                    print("Thus, the maximum point is the turning point (",turn_point,",",extreme,").")
                    if start == end:
                        print("There are two minimum points at (",c,",",start,"), and (",d,",",end,").") #The case that there are multiple minmum points in the interval.
                    if min(start, end) == start:
                        print("The minimum point is then (",c,",",start,").")
                    else:
                        print("The minimum point is then (",d,",",end,").")
             
            elif turn_point < c:   #Different reports based on both parabola and wether the turning point occurs before or after the closed interval:
                if self.coeffs[2] > 0:
                    print("This quadratic has a 'postitive' parabola, and the tunrning point is located before the start of the closed interval.")
                    print("Thus, the minimum point is the start of the closed interval (",c,",",start,").")
                    print("The maximum point is then the end of the closed interval (",d,",",end,").")
                
                elif self.coeffs[2] < 0:
                    print("This quadratic has a 'negative' parabola, and the tunrning point is located before the start of the closed interval.")
                    print("Thus, the maximum point is the start of the closed interval (",c,",",start,").")
                    print("The minimum point is then the end of the closed interval (",d,",",end,").")
                    
            elif turn_point > d: 
                if self.coeffs[2] > 0:
                    print("This quadratic has a 'postitive' parabola, and the tunrning point is located after the end of the closed interval.")
                    print("Thus, the maximum point is the start of the closed interval (",c,",",start,").")
                    print("The minimum point is then the end of the closed interval (",d,",",end,").")
                    
                elif self.coeffs[2] < 0:
                    print("This quadratic has a 'negative' parabola, and the tunrning point is located after the end of the closed interval.")
                    print("Thus, the minimum point is the start of the closed interval (",c,",",start,").")
                    print("The maximum point is then the end of the closed interval (",d,",",end,").")
                    
                    
        elif len(self.coeffs) == 2: #Case 2 for two coefficients (linear with non-zero x-coefficient).
            
            if self.coeffs[1] > 0:  #Different reports based on the coefficient of the x-term being positive or negative.
                print("This is a linear equation with the x-coefficent greater than 0.")
                print("Thus, the minimum point is the start of the closed interval (",c,",",start,").")
                print("The maximum point is then the end of the closed interval (",d,",",end,").")
                
            elif self.coeffs[1] < 0:
                print("This is a linear equation with the x-coefficent less than 0.")
                print("Thus, the maximum point is the start of the closed interval (",c,",",start,").")
                print("The minimum point is then the end of the closed interval (",d,",",end,").")
                
        elif len(self.coeffs) == 1:
            print("This is a constant function, so every point in the closed interval is a maximum and a minimum point, with value",self.coeffs[0],".")  #Same report for extreme values of a constant function each time.
            
                
                
#Question 1a)

F1 = Quadratic([-9,-4,2])
F1.roots()

#Question 1b)

F1.extrema(-4,-2) 

#Question 2a)

F2 = Quadratic([13,-4,2])
F2.roots()

#Question 2b)

F2.extrema(-1,3)

#Question 3a)

F3 = Quadratic([-5,2,0])
F3.roots()

The quadratic has two solutions to f(x) = 0; at x =  3.345207879911715 , and x =  -1.3452078799117149
This quadratic has a 'postitive' parabola, and the tunrning point is located after the end of the closed interval.
Thus, the maximum point is the start of the closed interval ( -4.0 , 39.0 ).
The minimum point is then the end of the closed interval ( -2.0 , 7.0 ).
The quadratic has two complex solutions to f(x) = 0; at x =  1.0 + 2.345207879911715 i , and x =  1.0 + -2.345207879911715 i
This quadratic has a 'postitive' parabola, and the tunrning point is located within the closed interval.
Thus, the minimum point is the turning point ( 1.0 , 11.0 ).
There are two maximum points at ( -1.0 , 19.0 ), and ( 3.0 , 19.0 ).
This linear equation has a solution to f(x) = 0 at x =  2.5


## Question 1
For question 1, were given the function 
$f_1(x) = 2x^2 - 4x - 9$, and were first asked to find the roots.
I first remove unwanted zeros to breakdown finding the roots into three cases, dependent on the number of coefficients. 
\
\
With 'D' and 'root_D' I find the discriminant, $b^2 - 4ac$, and root it, saving it using NumPy in case the result is a complex number. Splitting the result of the discriminant into three cases, I use the quadratic formula, $x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}$, to find the roots of the equation (using the discriminant value previously calculated) and print the roots cleanly, using .real and .imag.  For question 1a, the discriminant is $(-4)^2 - 4(2)(-9) = 88 > 0$, so we have (and correctly display) the two distinct real roots, $x_1 = 3.345207879911715$ and $x_2 = -1.3452078799117149$.
\
\
For question 1b, we analyse the extremal properties of the function $f_1$ giving the minimum and maximum points in a closed interval $[-4,-2]$. I first remove the unwanted zeros again to split finding the points into three cases.
\
\
Then using basic properties of quadratic functions, such as finding the tuning point (the $x$ value at $\frac{dy}{dx} = 0$), and the shape of the parabola (observing the $x^2$ coefficient), I split up the function into more cases and display all the extremal values, including the case for multiple minimum/maximum values. This displays the maximum point at the start of the closed interval, $x = -4$, resulting with the point $(-4,39)$, and the minimum point at the end of the closed interval, $x = -2$, resulting with the point $(-2,7)$.


## Question 2
For question 2, we are asked the same questions only now with the function $f_2(x) = 2x^2 - 4x + 13$, and with the closed interval $[-1,3]$ for part b.
\
\
For part a, I use the same 'roots' function and find two complex roots, $x = 1.0 + 2.345207879911715i$ and $x = 1.0 - 2.345207879911715i$.
\
\
For part b, I use the same 'extrema' function to find there's a minimum point at the turning point, $x = 1$, resulting with the point $(1,11)$, as well as two maximum points at the start and end of the closed interval, $x=-1$ and $x=3$ resulting with the two points $(-1,19)$ and $(3,19)$.

## Question 3
For question 3, we are asked only to find the roots of the function $f_3(x) = 2x - 5$.
\
\
Using the same roots function as before which makes use of the formula: $ax + b = 0 \Rightarrow x=-\frac{b}{a}, a\neq0$. This gives the one real root $x=2.5$.

